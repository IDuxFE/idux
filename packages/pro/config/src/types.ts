/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ProLocale } from '@idux/pro/locales'
import type { ProTableColumnIndexable } from '@idux/pro/table'

export interface ProGlobalConfig {
  common: ProCommonConfig
  locale: ProLocale

  table: ProTableConfig
  tree: ProTreeConfig
}

export type ProGlobalConfigKey = keyof ProGlobalConfig

export interface ProCommonConfig {
  prefixCls: string
}

export interface ProTableConfig {
  columnIndexable: Omit<ProTableColumnIndexable, 'type'>
}

export interface ProTreeConfig {
  clearIcon: string
  collapseIcon: [string, string]
}
