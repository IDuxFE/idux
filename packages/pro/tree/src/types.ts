/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { EmptyProps } from '@idux/components/empty'
import type { HeaderProps } from '@idux/components/header'
import type {
  TreeCheckStrategy,
  TreeCustomAdditional,
  TreeDragDropOptions,
  TreeDroppable,
  TreeNode,
  TreeNodeDisabled,
} from '@idux/components/tree'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const proTreeProps = {
  checkedKeys: { type: Array as PropType<VKey[]>, default: undefined },
  expandedKeys: { type: Array as PropType<VKey[]>, default: undefined },
  loadedKeys: { type: Array as PropType<VKey[]>, default: undefined },
  selectedKeys: { type: Array as PropType<VKey[]>, default: undefined },

  checkable: { type: Boolean, default: false },
  childrenKey: { type: String, default: undefined },
  cascaderStrategy: { type: String as PropType<TreeCheckStrategy>, default: 'all' },
  clearIcon: { type: String, default: undefined },
  customAdditional: { type: Object as PropType<TreeCustomAdditional>, default: undefined },
  collapsed: { type: Boolean, default: undefined },
  collapsedWidth: { type: Number, default: 44 },
  collapseIcon: { type: Array as PropType<string[]>, default: undefined },
  dataSource: { type: Array as PropType<TreeNode[]>, default: undefined },
  disabled: { type: Function as PropType<(node: TreeNode) => boolean | TreeNodeDisabled>, default: undefined },
  empty: { type: [String, Object] as PropType<string | EmptyProps>, default: undefined },
  draggable: { type: Boolean, default: false },
  draggableIcon: { type: String, default: undefined },
  droppable: { type: Function as PropType<TreeDroppable>, default: undefined },
  expandIcon: {
    type: [String, Array] as PropType<string | [string, string]>,
    default: () => ['minus-square', 'plus-square'],
  },
  getKey: { type: [String, Function] as PropType<string | ((data: TreeNode) => VKey)>, default: undefined },
  header: { type: [String, Object] as PropType<string | HeaderProps>, default: undefined },
  labelKey: { type: String, default: undefined },
  leafLineIcon: { type: String, default: undefined },
  loadChildren: { type: Function as PropType<(node: TreeNode) => Promise<TreeNode[]>>, default: undefined },
  showLine: { type: Boolean, default: true },
  searchValue: { type: String, default: undefined },
  searchable: { type: Boolean, default: true },
  searchFn: { type: Function as PropType<(node: TreeNode, searchValue?: string) => boolean>, default: undefined },
  selectable: { type: [Boolean, String] as PropType<boolean | 'multiple'>, default: true },
  placeholder: { type: String, default: undefined },
  virtual: { type: Boolean, default: false },

  // events
  onCheck: [Function, Array] as PropType<MaybeArray<<K = VKey>(checked: boolean, node: TreeNode<K>) => void>>,
  onClear: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onCollapsed: [Function, Array] as PropType<MaybeArray<(collapsed: boolean) => void>>,
  onDragstart: [Function, Array] as PropType<MaybeArray<<K = VKey>(options: TreeDragDropOptions<K>) => void>>,
  onDragend: [Function, Array] as PropType<MaybeArray<<K = VKey>(options: TreeDragDropOptions<K>) => void>>,
  onDragenter: [Function, Array] as PropType<MaybeArray<<K = VKey>(options: TreeDragDropOptions<K>) => void>>,
  onDragleave: [Function, Array] as PropType<MaybeArray<<K = VKey>(options: TreeDragDropOptions<K>) => void>>,
  onDragover: [Function, Array] as PropType<MaybeArray<<K = VKey>(options: TreeDragDropOptions<K>) => void>>,
  onDrop: [Function, Array] as PropType<MaybeArray<<K = VKey>(options: TreeDragDropOptions<K>) => void>>,
  onNodeClick: [Function, Array] as PropType<MaybeArray<<K = VKey>(evt: Event, node: TreeNode<K>) => void>>,
  onNodeContextmenu: [Function, Array] as PropType<MaybeArray<<K = VKey>(evt: Event, node: TreeNode<K>) => void>>,
  onLoaded: [Function, Array] as PropType<MaybeArray<<K = VKey>(loadedKeys: K[], node: TreeNode<K>) => void>>,
  onExpand: [Function, Array] as PropType<MaybeArray<<K = VKey>(expanded: boolean, node: TreeNode<K>) => void>>,
  onSelect: [Function, Array] as PropType<MaybeArray<<K = VKey>(selected: boolean, node: TreeNode<K>) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<<K = VKey>(startIndex: number, endIndex: number, visibleData: TreeNode<K>[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
  onSearch: [Function, Array] as PropType<MaybeArray<(searchText: string) => void>>,
  onExpandedChange: [Function, Array] as PropType<
    MaybeArray<<K = VKey>(expendedKeys: K[], expendedNodes: TreeNode<K>[]) => void>
  >,
  onCheckedChange: [Function, Array] as PropType<
    MaybeArray<<K = VKey>(checkedKeys: K[], checkedNodes: TreeNode<K>[]) => void>
  >,
  onSelectedChange: [Function, Array] as PropType<
    MaybeArray<<K = VKey>(selectedKeys: K[], selectedNodes: TreeNode<K>[]) => void>
  >,
  'onUpdate:collapsed': [Function, Array] as PropType<MaybeArray<(collapsed: boolean) => void>>,
  'onUpdate:checkedKeys': [Function, Array] as PropType<MaybeArray<<K = VKey>(keys: K[]) => void>>,
  'onUpdate:expandedKeys': [Function, Array] as PropType<MaybeArray<<K = VKey>(keys: K[]) => void>>,
  'onUpdate:loadedKeys': [Function, Array] as PropType<MaybeArray<<K = VKey>(keys: K[]) => void>>,
  'onUpdate:selectedKeys': [Function, Array] as PropType<MaybeArray<<K = VKey>(keys: K[]) => void>>,
} as const

export interface ProTreeBindings<K = VKey> {
  expandAll: () => void
  collapseAll: () => void
  /**
   * get node by it's key
   *
   * @param key
   * @returns node
   */
  getNode: (key: K) => TreeNode<K> | undefined
}

export type ProTreeProps = ExtractInnerPropTypes<typeof proTreeProps>
export type ProTreePublicProps = ExtractPublicPropTypes<typeof proTreeProps>
export type ProTreeComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ProTreePublicProps> & ProTreePublicProps,
  ProTreeBindings
>
export type ProTreeInstance = InstanceType<DefineComponent<ProTreeProps, ProTreeBindings>>
