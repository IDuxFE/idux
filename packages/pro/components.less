@import "./layout/style/index.less";
