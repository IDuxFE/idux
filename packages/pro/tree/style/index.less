@import '../../style/variable/index.less';
@import '../../../components/style/variable/index.less';
@import '../../style/mixins/reset.less';
@import './variables.less';

.@{pro-tree-prefix} {
  .reset-component();

  border: @pro-tree-border;
  padding: @pro-tree-padding;
  width: 100%;
  display: flex;
  flex-direction: column;
  overflow: hidden;
  transition: width var(--ix-motion-duration-medium) var(--ix-motion-ease-in-out);

  &-header-wrapper,
  &-search-wrapper {
    display: flex;
    align-items: center;
    gap: @pro-tree-header-search-wrapper-gap;
    padding: @pro-tree-header-search-wrapper-vertical-spacing @pro-tree-header-search-wrapper-horizontal-spacing;
  }

  &-search-wrapper {
    margin-bottom: @pro-tree-search-wrapper-margin-bottom;
    .@{input-prefix}-suffix {
      color: @pro-tree-search-input-suffix-color;
    }
  }

  &-header-wrapper {
    min-height: 40px;

    .@{icon-prefix} {
      font-size: @pro-tree-header-wrapper-icon-font-size;
      line-height: calc(@pro-tree-header-wrapper-icon-font-size + var(--ix-line-height-gutter));
      flex: auto;
      cursor: pointer;
      display: inline-block;
      position: relative;
      top: 2px;
      text-align: end;
      &:hover {
        color: @pro-tree-header-wrapper-icon-hover-color;
      }
    }

    & + .@{divider-prefix} {
      margin: @pro-tree-divider-vertical-spacing @pro-tree-divider-horizontal-spacing;
      width: calc(100% - @pro-tree-divider-horizontal-spacing * 2);
      min-width: calc(100% - @pro-tree-divider-horizontal-spacing * 2);
    }
  }

  .@{tree-prefix} {
    flex: 1;
    height: 0;
    padding: @pro-tree-content-vertical-spacing @pro-tree-content-horizontal-spacing;

    &.@{tree-prefix}-blocked {
      .@{tree-prefix}-node {
        padding: @pro-tree-node-padding;
      }
    }
  }

  &-collapsed {
    .@{tree-prefix} {
      .@{tree-prefix}-content {
        overflow: hidden;

        .cdk-virtual-scroll-holder {
          overflow: hidden !important;
        }
      }
    }
  }
  &-collapsed:not(&-animating) {
    .@{pro-tree-prefix} {
      &-header-wrapper {
        .@{header-prefix} {
          display: none;
        }

        justify-content: center;
      }

      &-search-wrapper {
        opacity: 0;
        visibility: hidden;
      }
    }

    .@{tree-prefix} {
      opacity: 0;
      visibility: hidden;
    }
  }
}
