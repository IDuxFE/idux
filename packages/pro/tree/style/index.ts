// style dependencies

import '@idux/components/icon/style'
import '@idux/components/button/style'
import '@idux/components/divider/style'

import './index.less'
