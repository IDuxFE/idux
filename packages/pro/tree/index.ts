/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ProTreeComponent } from './src/types'

import ProTree from './src/ProTree'

const IxProTree = ProTree as unknown as ProTreeComponent

export { IxProTree }

export type { ProTreeInstance, ProTreeComponent, ProTreePublicProps as ProTreeProps } from './src/types'

export { getThemeTokens as getProTreeThemeTokens } from './theme'

export type { ProTreeThemeTokens } from './theme'
