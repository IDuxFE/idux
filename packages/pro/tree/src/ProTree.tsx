/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass, normalizeStyle, ref } from 'vue'

import { isNil } from 'lodash-es'

import { callEmit, useControlledProp } from '@idux/cdk/utils'
import { ɵHeader } from '@idux/components/_private/header'
import { ɵInput } from '@idux/components/_private/input'
import { IxButton } from '@idux/components/button'
import { IxIcon } from '@idux/components/icon'
import { IxTree, type TreeInstance } from '@idux/components/tree'
import { useGlobalConfig } from '@idux/pro/config'

import { proTreeProps } from './types'

export default defineComponent({
  name: 'IxProTree',
  props: proTreeProps,
  setup(props, { expose, slots }) {
    const common = useGlobalConfig('common')
    const config = useGlobalConfig('proTree')

    const mergedPrefixCls = computed(() => `${common.prefixCls}-tree`)
    const expandIcon = computed(() => props.expandIcon ?? config.expandIcon)
    const mergedClearIcon = computed(() => props.clearIcon ?? config.clearIcon)

    const expandAllBtnStatus = ref(false)
    const treeRef = ref<TreeInstance>()

    const [searchValue, setSearchValue] = useControlledProp(props, 'searchValue', '')
    const [collapsed, setCollapsed] = useControlledProp(props, 'collapsed')

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value

      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-collapsed`]: collapsed.value,
      })
    })

    const style = computed(() => {
      return normalizeStyle({
        width: collapsed.value ? `${props.collapsedWidth}px` : null,
      })
    })

    const handleExpandAll = () => {
      const currStatus = expandAllBtnStatus.value
      expandAllBtnStatus.value = !currStatus
      expandAllBtnStatus.value ? treeRef.value?.collapseAll() : treeRef.value?.expandAll()
    }

    const handleCollapsed = () => {
      setCollapsed(!collapsed.value)
      callEmit(props.onCollapsed, !collapsed.value)
    }

    const handleInput = (evt: Event) => {
      setSearchValue((evt.target as HTMLInputElement).value)
      callEmit(props.onSearch, searchValue.value)
    }

    const handleClear = (evt: Event) => {
      setSearchValue('')
      callEmit(props.onClear, evt)
    }

    expose({
      blur: () => treeRef.value?.blur(),
      collapseAll: () => treeRef.value?.collapseAll(),
      expandAll: () => treeRef.value?.expandAll(),
      focus: () => treeRef.value?.focus(),
    })

    return () => {
      const prefixCls = mergedPrefixCls.value
      const treeProps = {
        checkedKeys: props.checkedKeys,
        expandedKeys: props.expandedKeys,
        selectedKeys: props.selectedKeys,
        loadedKeys: props.loadedKeys,

        blocked: true,
        cascade: true,
        childrenKey: props.childrenKey,
        checkable: props.checkable,
        checkStrategy: props.cascaderStrategy,
        customAdditional: props.customAdditional,
        dataSource: props.dataSource,
        disabled: props.disabled,
        expandIcon: expandIcon.value,
        empty: props.empty,
        getKey: props.getKey,
        labelKey: props.labelKey,
        leafLineIcon: props.leafLineIcon,
        showLine: props.showLine,
        searchValue: searchValue.value,
        selectable: props.selectable,
        searchFn: props.searchFn,
        virtual: props.virtual,

        loadChildren: props.loadChildren,
        onCheck: props.onCheck,
        onDragstart: props.onDragstart,
        onDragend: props.onDragend,
        onDragenter: props.onDragenter,
        onDragleave: props.onDragleave,
        onDragover: props.onDragover,
        onDrop: props.onDrop,
        onNodeClick: props.onNodeClick,
        onNodeContextmenu: props.onNodeContextmenu,
        onLoaded: props.onLoaded,
        onExpand: props.onExpand,
        onSelect: props.onSelect,
        onCheckedChange: props.onCheckedChange,
        onSelectedChange: props.onSelectedChange,
        onExpandedChange: props.onExpandedChange,
        onScroll: props.onScroll,
        onScrolledBottom: props.onScrolledBottom,
        onScrolledChange: props.onScrolledChange,
        'onUpdate:checkedKeys': props['onUpdate:checkedKeys'],
        'onUpdate:expandedKeys': props['onUpdate:expandedKeys'],
        'onUpdate:selectedKeys': props['onUpdate:selectedKeys'],
        'onUpdate:loadedKeys': props['onUpdate:loadedKeys'],
      }

      const treeSlots = {
        label: slots.label,
        prefix: slots.prefix,
        suffix: slots.suffix,
        leafLineIcon: slots.leafLineIcon,
        empty: slots.empty,
        expandIcon: slots.expandIcon,
      }
      return (
        <div class={classes.value} style={style.value}>
          <div class={`${prefixCls}-header-wrapper`}>
            <ɵHeader v-slots={slots} header={props.header} />
            {!isNil(collapsed.value) && <IxIcon name={collapsed.value ? 'right' : 'left'} onClick={handleCollapsed} />}
          </div>
          {props.searchable && (
            <div class={`${prefixCls}-search-wrapper`}>
              <IxButton
                size="xs"
                title={expandAllBtnStatus.value ? '展开全部' : '收起全部'}
                icon={expandAllBtnStatus.value ? 'tree-expand' : 'tree-unexpand'}
                onClick={handleExpandAll}
              />
              <ɵInput
                placeholder={props.placeholder}
                clearable
                clearIcon={mergedClearIcon.value}
                size="sm"
                suffix="search"
                value={searchValue.value}
                clearVisible={!!searchValue.value}
                onInput={handleInput}
                onClear={handleClear}
              />
            </div>
          )}
          <IxTree ref={treeRef} v-slots={treeSlots} {...treeProps} />
        </div>
      )
    }
  },
})
