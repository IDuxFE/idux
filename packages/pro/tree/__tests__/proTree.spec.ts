import { MountingOptions, mount } from '@vue/test-utils'

import { renderWork, wait } from '@tests'

import { type TreeNode } from '@idux/components/tree'

import ProTree from '../src/ProTree'
import { ProTreeProps } from '../src/types'

const defaultDataSource = [
  {
    label: 'Node 0',
    key: '0',
    children: [
      {
        label: 'Node 0-0',
        key: '0-0',
      },
      {
        label: 'Node 0-1',
        key: '0-1',
      },
      {
        label: 'Node 0-2',
        key: '0-2',
      },
    ],
  },
]

const expandedKeys = ['0']

describe('ProTree', () => {
  const ProTreeMount = (options?: MountingOptions<Partial<ProTreeProps>>) => {
    const { props, ...rest } = options || {}
    return mount(ProTree, {
      ...rest,
      props: { dataSource: defaultDataSource, expandedKeys, ...props },
      attachTo: 'body',
    })
  }

  renderWork<ProTreeProps>(ProTree, {
    props: { dataSource: defaultDataSource },
  })

  test('checkable work', async () => {
    const wrapper = ProTreeMount()

    expect(wrapper.find('.ix-tree-node-checkbox').exists()).toBe(false)

    await wrapper.setProps({ checkable: true })

    expect(wrapper.find('.ix-tree-node-checkbox').exists()).toBe(true)
  })

  test('searchable work', async () => {
    const wrapper = ProTreeMount()

    expect(wrapper.find('.ix-pro-tree-search-wrapper').exists()).toBe(true)
    expect(wrapper.find('.ix-divider').exists()).toBe(false)

    await wrapper.setProps({ searchable: false })

    expect(wrapper.find('.ix-tree-node-checkbox').exists()).toBe(false)
    expect(wrapper.find('.ix-divider').exists()).toBe(true)

    await wrapper.setProps({ collapsed: true })

    expect(wrapper.find('.ix-divider').exists()).toBe(false)

    await wrapper.setProps({ collapsed: false })

    expect(wrapper.find('.ix-divider').exists()).toBe(true)
  })

  test(':checkedKeys.sync work', async () => {
    const onUpdateCheckedKeys = vi.fn()
    const onCheckedChange = vi.fn()

    const wrapper = ProTreeMount({
      props: {
        checkable: true,
        checkedKeys: ['0'],
        'onUpdate:checkedKeys': onUpdateCheckedKeys,
        onCheckedChange,
      },
    })

    const allNodes = wrapper.findAll('.ix-tree-node')

    allNodes.forEach(node => {
      expect(node.find('.ix-checkbox-checked').exists()).toBe(true)
    })

    // 0-0, unchecked
    await allNodes[1].find('input').setValue(false)

    expect(onUpdateCheckedKeys).toBeCalledWith(['0-1', '0-2'])
    expect(onCheckedChange).toBeCalledWith(
      ['0-1', '0-2'],
      [
        {
          key: '0-1',
          label: 'Node 0-1',
        },
        {
          key: '0-2',
          label: 'Node 0-2',
        },
      ],
    )

    await wrapper.setProps({ checkedKeys: ['0-0'] })

    // 0 indeterminate
    expect(allNodes[0].find('.ix-checkbox-indeterminate').exists()).toBe(true)
  })

  test(':expandedKeys.sync work', async () => {
    const onUpdateExpandedKeys = vi.fn()
    const wrapper = ProTreeMount({
      props: {
        dataSource: [
          {
            label: 'Node 0',
            key: '0',
            children: [
              {
                label: 'Node 0-0',
                key: '0-0',
                children: [
                  {
                    label: 'Node 0-0-0',
                    key: '0-0-0',
                  },
                ],
              },
              {
                label: 'Node 0-1',
                key: '0-1',
              },
              {
                label: 'Node 0-2',
                key: '0-2',
              },
            ],
          },
        ],
        expandedKeys: ['0'],
        'onUpdate:expandedKeys': onUpdateExpandedKeys,
      },
    })

    const allNodes = wrapper.findAll('.ix-tree-node')

    expect(allNodes[0].classes()).toContain('ix-tree-node-expanded')

    await wrapper.setProps({ expandedKeys: ['0', '0-0'] })

    expect(allNodes[0].classes()).toContain('ix-tree-node-expanded')
    expect(allNodes[1].classes()).toContain('ix-tree-node-expanded')

    // 0-0
    await allNodes[1].find('.ix-tree-node-expand').trigger('click')

    expect(onUpdateExpandedKeys).toBeCalledWith(['0'])

    await wrapper.setProps({ expandedKeys: ['0'] })

    expect(allNodes[0].classes()).toContain('ix-tree-node-expanded')
    expect(allNodes[1].classes()).not.toContain('ix-tree-node-expanded')

    // 0-0
    await allNodes[1].find('.ix-tree-node-expand').trigger('click')

    expect(onUpdateExpandedKeys).toBeCalledWith(['0', '0-0'])

    await wrapper.setProps({ expandedKeys: ['0', '0-0'] })

    expect(allNodes[0].classes()).toContain('ix-tree-node-expanded')
    expect(allNodes[1].classes()).toContain('ix-tree-node-expanded')
  })

  test(':selectedKeys.sync work', async () => {
    const onUpdateSelectedKeys = vi.fn()
    const wrapper = ProTreeMount({
      props: { selectedKeys: ['0'], 'onUpdate:selectedKeys': onUpdateSelectedKeys },
    })

    const allNodes = wrapper.findAll('.ix-tree-node')

    expect(allNodes[0].classes()).toContain('ix-tree-node-selected')

    await wrapper.setProps({ selectedKeys: ['0', '0-0'] })

    expect(allNodes[0].classes()).toContain('ix-tree-node-selected')
    expect(allNodes[1].classes()).toContain('ix-tree-node-selected')

    // 0-0
    await allNodes[1].find('.ix-tree-node-content').trigger('click')

    expect(onUpdateSelectedKeys).toBeCalledWith([])

    await wrapper.setProps({ selectedKeys: [] })

    expect(allNodes[0].classes()).not.toContain('ix-tree-node-selected')
    expect(allNodes[1].classes()).not.toContain('ix-tree-node-selected')

    // 0-0
    await allNodes[1].find('.ix-tree-node-content').trigger('click')

    expect(onUpdateSelectedKeys).toBeCalledWith(['0-0'])

    await wrapper.setProps({ selectedKeys: ['0-0'] })

    expect(allNodes[0].classes()).not.toContain('ix-tree-node-selected')
    expect(allNodes[1].classes()).toContain('ix-tree-node-selected')
  })

  test('loadedKeys work', async () => {
    const onUpdateLoadedKeys = vi.fn()
    const loadChildren = (node: TreeNode) => {
      return new Promise<TreeNode[]>(resolve => {
        setTimeout(() => {
          const parentKey = node.key as string
          const children = [
            { label: `Child ${parentKey}-0 `, key: `${parentKey}-0` },
            { label: `Child ${parentKey}-1 `, key: `${parentKey}-1` },
          ]
          resolve(children)
        }, 50)
      })
    }

    const wrapper = ProTreeMount({
      props: {
        dataSource: [
          { key: '0', label: '0' },
          { key: '1', label: '1', isLeaf: true },
        ],
        expandedKeys: undefined,
        checkedKeys: undefined,
        selectedKeys: undefined,
        loadChildren,
        loadedKeys: ['0'],
        'onUpdate:loadedKeys': onUpdateLoadedKeys,
      },
    })

    let allNodes = wrapper.findAll('.ix-tree-node')

    expect(allNodes.length).toBe(2)

    await allNodes[0].find('.ix-tree-node-expand').trigger('click')

    expect(allNodes[0].find('.ix-tree-node-expand').find('.ix-icon-loading').exists()).toBe(false)

    await wait(50)

    allNodes = wrapper.findAll('.ix-tree-node')

    expect(allNodes.length).toBe(2)

    await wrapper.setProps({ loadedKeys: [] })

    await allNodes[0].find('.ix-tree-node-expand').trigger('click')

    expect(allNodes[0].find('.ix-tree-node-expand').find('.ix-icon-loading').exists()).toBe(true)

    await wait(50)

    allNodes = wrapper.findAll('.ix-tree-node')

    expect(allNodes.length).toBe(4)
  })

  test(':collapsed.sync work', async () => {
    const onUpdateCollapsed = vi.fn()
    const onCollapsed = vi.fn()

    const wrapper = ProTreeMount({
      props: {
        'onUpdate:collapsed': onUpdateCollapsed,
        onCollapsed,
      },
    })

    expect(wrapper.find('.ix-pro-tree-collapsed-icon').exists()).toBe(false)

    await wrapper.setProps({ collapsed: false })

    expect(wrapper.find('.ix-pro-tree-collapsed-icon').exists()).toBe(true)
    expect(wrapper.find('.ix-icon-collapse').exists()).toBe(true)

    await wrapper.setProps({ collapsed: true })

    expect(wrapper.find('.ix-icon-uncollapse').exists()).toBe(true)

    await wrapper.find('.ix-pro-tree-collapsed-icon').trigger('click')

    expect(onUpdateCollapsed).toBeCalledWith(false)
    expect(onCollapsed).toBeCalledWith(false)
  })

  test('collapsedWidth work', async () => {
    const wrapper = ProTreeMount({
      props: {
        collapsed: true,
      },
    })

    expect(wrapper.attributes('style')).toContain('width: 44px')

    await wrapper.setProps({ collapsedWidth: 100 })

    expect(wrapper.attributes('style')).toContain('width: 100px')
  })

  test('placeholder work', async () => {
    const wrapper = ProTreeMount({
      props: {
        placeholder: 'placeholder',
      },
    })

    expect(wrapper.find('.ix-pro-tree-search-wrapper .ix-input-inner').attributes('placeholder')).toBe('placeholder')

    await wrapper.setProps({ placeholder: 'temp' })

    expect(wrapper.find('.ix-pro-tree-search-wrapper .ix-input-inner').attributes('placeholder')).toBe('temp')
  })

  test('header work', async () => {
    const wrapper = ProTreeMount({
      props: {
        header: 'title',
      },
    })

    expect(wrapper.find('.ix-header-title').text()).toBe('title')

    await wrapper.setProps({
      header: {
        title: 'tmp',
        suffix: 'setting',
      },
    })

    expect(wrapper.find('.ix-header-title').text()).toBe('tmp')
    expect(wrapper.find('.ix-header-suffix .ix-icon-setting').exists()).toBe(true)
  })

  test('header slot work', async () => {
    const wrapper = ProTreeMount({
      props: {
        header: 'title',
      },
      slots: {
        header: 'header',
      },
    })

    expect(wrapper.find('.ix-pro-tree-header-wrapper').text()).toBe('header')
    expect(wrapper.find('.ix-header').exists()).toBe(false)
  })

  test('clearIcon work', async () => {
    const wrapper = ProTreeMount({
      props: {
        searchValue: 'searchValue',
      },
    })

    expect(wrapper.find('.ix-input-clear .ix-icon-close-circle').exists()).toBe(true)

    await wrapper.setProps({
      clearIcon: 'setting',
    })

    expect(wrapper.find('.ix-input-clear .ix-icon-setting').exists()).toBe(true)
  })

  test('collapseIcon work', async () => {
    const wrapper = ProTreeMount()

    expect(wrapper.find('.ix-pro-tree-collapsed-icon').exists()).toBe(false)

    await wrapper.setProps({
      collapsed: false,
    })

    expect(wrapper.find('.ix-pro-tree-collapsed-icon').exists()).toBe(true)
    expect(wrapper.find('.ix-icon-collapse').exists()).toBe(true)

    await wrapper.setProps({
      collapseIcon: ['left', 'right'],
    })

    expect(wrapper.find('.ix-icon-left').exists()).toBe(true)
    expect(wrapper.find('.ix-icon-collapse').exists()).toBe(false)

    await wrapper.setProps({
      collapsed: true,
    })

    expect(wrapper.find('.ix-icon-right').exists()).toBe(true)
    expect(wrapper.find('.ix-icon-uncollapse').exists()).toBe(false)
  })

  test('searchValue.sync work', async () => {
    const onSearch = vi.fn()
    const onUpdateSearch = vi.fn()
    const wrapper = ProTreeMount({
      props: {
        searchValue: 'Node 0-0',
        onSearch,
        'onUpdate:searchValue': onUpdateSearch,
      },
    })

    const input = wrapper.find('.ix-pro-tree-search-wrapper .ix-input-inner')

    expect(!!wrapper.findAll('.ix-tree-node')[1]?.find('.ix-tree-node-content-label-highlight').exists()).toBe(true)

    await wrapper.setProps({
      searchValue: 'tmp',
    })

    expect(!!wrapper.findAll('.ix-tree-node')[1]?.find('.ix-tree-node-content-label-highlight').exists()).toBe(false)

    await input.setValue('search text')

    expect(onSearch).toBeCalledWith('search text')
    expect(onUpdateSearch).toBeCalledWith('search text')
  })

  test('expandAll and collapseAll work', async () => {
    const onUpdateExpandedKeys = vi.fn()
    const wrapper = ProTreeMount({
      props: {
        expandedKeys: ['0'],
        'onUpdate:expandedKeys': onUpdateExpandedKeys,
      },
    })

    await wrapper.vm.collapseAll()

    expect(onUpdateExpandedKeys).toBeCalledWith([])

    await wrapper.vm.expandAll()

    expect(onUpdateExpandedKeys).toBeCalledWith(['0'])
  })
  test('getNode work', async () => {
    const onUpdateExpandedKeys = vi.fn()
    const wrapper = ProTreeMount({
      props: {
        dataSource: defaultDataSource,
        expandedKeys: ['0'],
        'onUpdate:expandedKeys': onUpdateExpandedKeys,
      },
    })

    expect(wrapper.vm.getNode('0')).toMatchObject(defaultDataSource[0])

    expect(wrapper.vm.getNode('0-1')).toMatchObject(defaultDataSource[0].children[1])

    expect(wrapper.vm.getNode('4')).toBe(undefined)
  })
})
