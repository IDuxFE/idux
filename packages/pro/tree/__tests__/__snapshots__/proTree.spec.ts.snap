// Vitest Snapshot v1

exports[`ProTree > render work 1`] = `
"<div class=\\"ix-pro-tree\\">
  <div class=\\"ix-pro-tree-header-wrapper\\">
    <!---->
    <!---->
  </div>
  <div class=\\"ix-pro-tree-search-wrapper\\"><button class=\\"ix-button ix-button-icon-only ix-button-default ix-button-square ix-button-xs\\" type=\\"button\\" title=\\"收起全部\\"><i class=\\"ix-icon ix-icon-tree-unexpand\\" role=\\"img\\" aria-label=\\"tree-unexpand\\"></i>
      <div aria-hidden=\\"true\\" class=\\"ix-wave\\"></div>
    </button><span class=\\"ix-input ix-input-sm ix-input-clearable ix-input-with-suffix\\"><!----><input class=\\"ix-input-inner\\"><span class=\\"ix-input-suffix\\"><i class=\\"ix-icon ix-icon-search\\" role=\\"img\\" aria-label=\\"search\\"></i></span><span class=\\"ix-input-clear\\"><i class=\\"ix-icon ix-icon-close-circle\\" role=\\"img\\" aria-label=\\"close-circle\\"></i></span></span></div>
  <div class=\\"ix-tree ix-tree-blocked ix-tree-auto-height ix-tree-show-line\\" role=\\"tree\\">
    <!----><input style=\\"width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;\\" tabindex=\\"0\\" aria-label=\\"for screen reader\\">
    <div class=\\"ix-tree-content\\">
      <div class=\\"ix-tree-content-inner\\">
        <div class=\\"ix-tree-node ix-tree-node-last\\" aria-label=\\"Node 0\\" aria-selected=\\"false\\" title=\\"Node 0\\"><span aria-hidden=\\"true\\" class=\\"ix-tree-node-indent\\"></span>
          <!----><span class=\\"ix-tree-node-expand\\"><!----><i class=\\"ix-icon ix-icon-plus-square\\" style=\\"transform: rotate(0deg);\\" role=\\"img\\" aria-label=\\"plus-square\\"></i></span>
          <!----><span class=\\"ix-tree-node-content\\"><!----><span class=\\"ix-tree-node-content-label\\">Node 0</span>
          <!----></span>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
